RavenDB-20689 - Improve the performance of a new index when it includes filtering

grisha-kotler opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20689/Slow-performance-of-the-Bloom-Filters

### Additional description

Improve the performance of a new index when it includes filtering.

### Type of change

- Optimization

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

grisha-kotler committed:
RavenDB-20689 - improve the performance when the index includes filtering

arekpalinski commented on src/Raven.Server/Documents/Indexes/MapIndexBase.cs:
```diff
@@ -122,15 +122,15 @@ private int UpdateIndexEntriesCorax(IndexItem indexItem, IEnumerable mapResults,
+            if (indexItem.SkipLuceneDelete == false)
             {
-                mustDelete = _filters.Add(indexItem.LowerId) == false;
+                using (_stats.BloomStats.Start())
```
Please note that now you're including the duration of  `writer.Value.Delete(indexItem.LowerId, stats);` as bloom filter processing

grisha-kotler replied:
Correct. Fixed.

ppekrol commented on src/Raven.Server/Documents/Indexes/MapIndexBase.cs:
```diff
@@ -139,6 +139,14 @@ private int UpdateIndexEntriesLucene(IndexItem indexItem, IEnumerable mapResults
                 numberOfOutputs++;
             }
 
+            if (indexItem.SkipLuceneDelete && numberOfOutputs > 0)
```
this approach I think can be optimized further (please correct me if I'm wrong) cc @ayende 

if we would do Contains first then we would know that the set does not contain that Id, which means that if numberOfOutputs is still 0, we might skip the Add?

Then it would work not only for first run, but always?

ayende replied:
How common is that? You'll need:
* New document
* Filtering

no?

ppekrol replied:
What @grisha-kotler did here, only helps for 1st run, if I will update the document, even if it is filtered out anyway, it will hit the bloom on 2nd and subsequent runs

ayende replied:
Okay, so the idea is that if it isn't there and was filtered, we can skip.
Yes, makes sense

grisha-kotler replied:
Done

grisha-kotler committed:
RavenDB-20689 - calculate only the time that it took us to update the Bloom Filters

grisha-kotler commented on src/Raven.Server/Documents/Indexes/MapIndexBase.cs:
```diff
@@ -139,6 +139,14 @@ private int UpdateIndexEntriesLucene(IndexItem indexItem, IEnumerable mapResults
+                // we skip deleting from lucene for the initial indexing run
+                // but we must update the bloom filters if we actually indexed that document
+                using (_stats.BloomStats.Start())
+                    _filters.Add(indexItem.LowerId);
```
I think that we can optimize this one as well :)
`Add` is checking if the id exists on all existing filters.
At the first indexing run, we know that it isn't possible. So we can just directly add it.
@ayende @ppekrol

ppekrol replied:
Check my other comment, it should solve your issue

grisha-kotler replied:
Done

grisha-kotler committed:
RavenDB-20689 - optimize the usage of bloom filters for 2 cases:
- on the initial index run
- when we have nothing to index but still need might not need to delete from the index
